plugins/ltac/extratactics.mlg

```diff
@@ -1119,3 +1119,52 @@ let tclOPTIMIZE_HEAP =
 TACTIC EXTEND optimize_heap
 | [ "optimize_heap" ] -> { tclOPTIMIZE_HEAP }
 END
+
+(** tactic analogous to "Strategy" *)
+
+{
+
+let with_set_strategy lvl ql k =
+  let glob_key r =
+    match r with
+    | GlobRef.ConstRef sp -> ConstKey sp
+    | GlobRef.VarRef id -> VarKey id
+    | _ -> user_err Pp.(str
+        "cannot set an inductive type or a constructor as transparent") in
+  Proofview.tclENV >>= fun env ->
+  let orig_kl = List.map (fun k ->
+      (k, Conv_oracle.get_strategy (Environ.oracle env) k))
+      (List.map glob_key ql) in
+  (* because the global env might be desynchronized from the proof-local
+     env, and we need to update the global env to have this tactic play
+     nicely with abstract *)
+  let orig_kl_global = List.map (fun k ->
+      (k, Conv_oracle.get_strategy (Environ.oracle (Global.env ())) k))
+      (List.map glob_key ql) in
+  let env = List.fold_left (fun env q ->
+    Environ.set_oracle env
+      (Conv_oracle.set_strategy (Environ.oracle env) (glob_key q) lvl)) env ql in
+  Proofview.Goal.enter begin
+    fun _ ->
+      List.iter (fun q -> Global.set_strategy (glob_key q) lvl) ql;
```

Comment:
I'm very sorry but I'm working hard on removing access to `Global` from tactics so I'd rather no see new calls added.

Reply:
When the call in `abstract` is removed, the call in this tactic can be removed at the same time.  The effect should not be noticeable except via the `abstract` tactic (and other tactics that inspect the global env); I added the calls to the global env just to support `abstract` (and kept them in a separate commit, so it should be easy to revert the changes that commit makes to this file). (And note that the code to set the proof-local env is already in place, so you won't even need to replace the global calls with anything once `abstract` no longer relies on the global env.)  I can add a comment to this effect.  Would this be sufficient?  Or am I missing something about why you want to not see new calls added?

The alternative would be to drop the second commit from this PR (but then we'll have to document that the interaction with `abstract` is broken, or else I can just open an issue about it).

Reply:
Thanks for the answer @JasonGross , let me think about it.

Reply:
Overall I'd say that either adding the comment or leaving abstract broken should work; my fear with the first option that we suddenly get other parts of the system to depend on the global state so migration to a purely functional tactic state becomes harder.

Having tactics to behave purely functional is IMO extremely important as that will allow Coq to profit from the upcoming OCaml multicore and that is feature we _all_ want to have ASAP :) but indeed, there are tradeoffs.

Reply:
I've updated the code and added a bunch of `TODO`s with instructions on what to remove once `abstract` no longer depends on `Global`.

Indeed, I support moving to a purely functional tactic state (though I do think there should be ways to update the document state from tactics, such as `abstract` and #11029, but it seems better to implement this in a functional way).

Reply:
Your point about abstract is an interesting one; cc #9146 ; I am not sure the view where abstract does modify document state is the one we want; I'd rather see core calculus let extended such that:
- `let x := foo in y` does allow to specify a transparency in the term; this is enforced for everything.
- `let x := foo in y` does have a trusted mode where `foo` is only type-checked once.

I fact I did a prototype of this last one as it was satisfactory, only problem is that it wasn't easy to design an API where this invariant can be enforced.

Reply:
That sounds like it is in the right direction, and I have thoughts about it, but I'll hold them for a venue that's actually about `abstract` rather than sidetracking this PR more.

Reply:
It would great to hear your thoughts on abstract tho, a rework of how it works is likely to happen soon so any input is very welcome.

Regarding this PR, my preference sits in the option of "leaving abstract broken" as I'm unsure how this would play with future plans otherwise, but I'll let @ppedrot to have the final say on this, thus his review should override mine.

Reply:
> Having tactics to behave purely functional is IMO extremely important as that will allow Coq to profit from the upcoming OCaml multicore

... except that tactics are the least likely part of the system to take advantage of this feature. By definition they are a state monad where the state is a heap of blobs, so the kind of stuff really hard to run in parallel. I am much more confident that we can put multicore to good use in the kernel, which happens to be already purely functional (mostly).

Reply:
Not at all @ppedrot ; once tactics are functional you can hope to run the bits between `Proof.` and `Qed.` in a separate thread; with some other minor tricks you can indeed hope to _build_ opaque proofs in their own thread _à la_ isabelle.

Reply:
Ah, indeed. I thought you wanted to have fine-grained parallelism in tactics à la `par:`.

Reply:
Oh good point, yesterday I was thinking about `par:` as is one of the few bits handled in code I'm replacing that I was not sure how to handle; not sure what should be done here.